ElectricCar.cpp

```diff
@@ -4,13 +4,48 @@
 ElectricCar::ElectricCar(ElectricEngine* engine)
     : engine_(engine)
 {
-    std::cout << __FUNCTION__ << std::endl;
+    std::cout << "* " << __FUNCTION__ << std::endl;
 }
 
-ElectricCar::~ElectricCar()       { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::turnLeft()      { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::turnRight()     { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::brake()         { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::accelerate(int) { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::charge()        { std::cout << __FUNCTION__ << std::endl; }
+ElectricCar::~ElectricCar() {
+    std::cout << "* " << __FUNCTION__ << std::endl;
+    delete engine_;
+}
+
+void ElectricCar::charge() { 
+    std::cout << __FUNCTION__;
+    if (chargeLevel_ < MAX_CAPACITY) {
+        chargeLevel_ = MAX_CAPACITY;
+        std::cout << " : recharged" << std::endl;
+    } else {
+        std::cerr << " : battery already full" << std::endl;
+    }
+}
+
+size_t ElectricCar::getChargeLevel() {
+    return chargeLevel_;
+}
 
+void ElectricCar::setEngine(ElectricEngine* engine) {
+    std::cout << __FUNCTION__;
+    if (!getSpeed()) {
+        std::cout << " : engine changed" << std::endl;
+        delete engine_;
+        engine_ = engine;
+        std::cout << "@ electric power: " << engine_->getPower() << std::endl;
+    } else {
+        std::cout << " : not possible unless stopped" << std::endl;
+        delete engine;
+    }
+}
+void ElectricCar::setEngine(PetrolEngine*) {}
+void ElectricCar::setEngine(PetrolEngine*, ElectricEngine*) {}
```

Comment:
Good idea, but use `= delete` from C++11 in header files :)

Reply:
Cannot be deleted. When you inherit from a class that has pure virtual functions, you MUST implement those functions. I know it's a major design fault, and needs some design refactoring. What am I to do then? Make them private at least I guess.

Reply:
Then don't put it into interface. It's breaking Liskov Substitution Principle from SOLID rules. Or at least don't make them pure virtual.

Reply:
I will refactor tomorrow then.

Reply:
Pure virtual methods setEngine() removed from interface and used appropriately in PetrolCar, ElectricCar & HybridCar classes. The disadvantage is that the engine cannot be changed via a pointer to the interface. I was guided by the 'Obiektowosc #4' lesson where we used 'Cars' interface to solve the diamond problem. Hope that's good enough.